File: Golang/p1_2Sum/2Sum.go
package main

import "fmt"

func twoSum(nums []int, target int) []int {
	hashMap := make(map[int]int)
	//Index loops
	for i := 0; i < len(nums); i++ {
		//Find remain each loop
		remain := target - nums[i]
		//If hashMap have remain val then return value of key
		if _, ok := hashMap[remain]; ok {
			return []int{hashMap[remain], i}
		}
		//Else must insert key,val into hashMap
		hashMap[nums[i]] = i
	}
	return nil
}
func main() {
	// Example inputs
	nums1 := []int{2, 7, 11, 15}
	target1 := 9

	nums2 := []int{3, 2, 4}
	target2 := 6

	nums3 := []int{3, 3}
	target3 := 6

	// Test twoSum with the inputs
	fmt.Println("Test Case 1:")
	fmt.Printf("Input: nums = %v, target = %d\n", nums1, target1)
	fmt.Printf("Output: %v\n", twoSum(nums1, target1))

	fmt.Println("Test Case 2:")
	fmt.Printf("Input: nums = %v, target = %d\n", nums2, target2)
	fmt.Printf("Output: %v\n", twoSum(nums2, target2))

	fmt.Println("Test Case 3:")
	fmt.Printf("Input: nums = %v, target = %d\n", nums3, target3)
	fmt.Printf("Output: %v\n", twoSum(nums3, target3))
}

File: Python/p1_2Sum/2Sum.py
#Given an array of integers nums and an integer target,
# then return indices of the two numbers such that they add up to target.

#Ex:nums = [2,7,11,15], target = 9
#Output : [0,1]
#Explanation:  Because nums[0] + nums[1] == 9, we return [0, 1]

#Intuitation:Using hashmap 
from typing import List

class Solution :
    #Assume we have a list [2,7,11,15]
     def twoSum(self, nums: List[int], target: int) -> List[int]:
         map = {}
         
         for index,value in enumerate(nums):
             #Assume target = 9 
             #In the first loop 
             #-> remain = 9 -2 =7
             remain = target-value
             #First checking will check remain = 7 
             #Cause there's no ele in map then it must insert first
             #After insert , check again 
             if remain in map :
                 #And return the answer
                 return [map[remain],index]
             #Insert a key and a value : 
             #K-> Key(value) : 2 , Index = 0
             map[value] = index

#Test function 
def test_two_sum():
    solution = Solution()

    # Test case 1
    nums1 = [2, 7, 11, 15]
    target1 = 9
    expected_output1 = [0, 1]
    assert solution.twoSum(nums1, target1) == expected_output1, f"Test case 1 failed: {solution.twoSum(nums1, target1)}"

    # Test case 2
    nums2 = [3, 2, 4]
    target2 = 6
    expected_output2 = [1, 2]
    assert solution.twoSum(nums2, target2) == expected_output2, f"Test case 2 failed: {solution.twoSum(nums2, target2)}"

    # Test case 3
    nums3 = [3, 3]
    target3 = 6
    expected_output3 = [0, 1]
    assert solution.twoSum(nums3, target3) == expected_output3, f"Test case 3 failed: {solution.twoSum(nums3, target3)}"

    print("All test cases passed!")

# Run the test function
test_two_sum()

             